### DISCUSSION TITLE
lenght, time and energy base scale in Crow app

### DISCUSSION THREAD

**Discussion by Giovar6** [idaholab/moose#16757] [Q&A Modules: Phase field] [answered]
Dear MOOSE users,

I'm here again with a problem with the Crow app, developed by @SudiptaBiswas. This time is more a conceptual problem.
In the material section of the input file, when calculating the mobility properties and the simulation coefficients, in the Crow application are specified some coefficients to non-dimensionalize the material properties.
To explain better what I'm referring to, I report an example of an input file developed for the Crow app. (I quote only the modules  of the input file important to clarify what I am referring to)

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 80
  ny = 40
  nz = 0
  xmin = 0.0
  xmax = 400.0
  ymin = 0.0
  ymax = 200.0
  zmax = 0
  uniform_refine = 1
  elem_type = QUAD4
[]

[Materials]
  [./free_energy]
    type = SinteringFreeEnergy
    block = 0
    c = c
    v = 'gr0 gr1'
    #A = A
    #B = B
    #f_name = S
    derivative_order = 2
    #outputs = console
  [../]
  [./coeff]
    type = SinteringCoefficients
    block = 0
    T = 800.0
    int_width = 20
    GBmob0 = 2.5e-6
    Qgbm = 0.23
    surface_energy = 1.03e19
    GB_energy = 4.42e18
    length_scale = 1e-08
    time_scale = 0.1
    energy_scale = 100.0
    energy_unit = eV
    outputs = exodus
  [../]
  [./mob]
    type = SinteringMtrxMobility
    T = 800.0
    int_width = 20
    Qv = 2.19
    Qvc = 2.19
    Qgb = 0.75
    Qs = 0.9
    Dgb0 = 1.95e-8
    Dsurf0 = 2.6e-5
    Dvap0 = 7.8e-5
    Dvol0 = 7.8e-5
    c = c
    v = 'gr0 gr1'
    Vm = 1.182e-29
    length_scale = 1e-08
    time_scale = 0.1
    bulkindex = 1.0
    surfindex = 1.0
    gbindex = 1.0
    #prefactor = 0.1
    outputs = exodus
  [../]
[]

[Executioner]
  Preconditioned JFNK (default)
  type = Transient
  scheme = BDF2
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap -sub_pc_factor_shift_type'
  petsc_options_value = 'asm         31   preonly   lu      1 nonzero'
  #petsc_options = '-snes_converged_reason'
  l_max_its = 20
  nl_max_its = 20
  nl_abs_tol = 1e-08
  nl_rel_tol = 1e-06
  l_tol = 1e-04
  end_time = 5000
  dt = 0.01
  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.01
    growth_factor = 1.2
  [../]
[]

[ICs]
  [./ic_gr1]
    int_width = 20.0
    x1 = 250.0
    y1 = 100.0
    radius = 74.0
    outvalue = 0.0
    variable = gr1
    invalue = 1.0
    type = SmoothCircleIC
  [../]
  [./multip]
    x_positions = '100.0 250.0'
    int_width = 20.0
    z_positions = '0 0'
    y_positions = '100.0 100.0 '
    radii = '74.0 74.0'
    3D_spheres = false
    outvalue = 0.001
    variable = c
    invalue = 0.999
    type = SpecifiedSmoothCircleIC
    block = 0
  [../]
  [./ic_gr0]
    int_width = 20.0
    x1 = 100.0
    y1 = 100.0
    radius = 74.0
    outvalue = 0.0
    variable = gr0
    invalue = 1.0
    type = SmoothCircleIC
  [../]
[]

In this case are are adopted as length scale 1e-08, time scale 0.1 and energy scale 100. The problem is that i cannot understand how this exact values are identified. Why I cannot use for example 1e-06?
I have tried to perform simulations with these values, but the simulation did not converged. So I understand that is important to specify the correct values, but how?

Thank you in advance to anyone will give me any suggestion or advice.
Giovanni Rizza

**Comment by aeslaughter** [comment]
@SudiptaBiswas This is super old, but curious if you had any advice.

**Comment by SudiptaBiswas** [comment, marked as answer]
@Giovar6 The scales are chosen such that the nondimensional mobility values are close to unity. This is done for computational benefits. This is generic for phase-field models, nothing specific to Crow. Feel free to use different scales that make sense for your simulations.

**Reply by Giovar6** [reply]
Hello @SudiptaBiswas  and thank you for your reply.
Thank you for the suggestions and the help. I will try to identify the best scales also with some further literature research.

**Reply by HPL001** [reply]
@SudiptaBiswas Hellow, I am a PhD student studying MOOSE, I am very interested in your application Crow, thank you so much for providing such excellent code. I from this web site (https://github.com/SudiptaBiswas/Crow) to download the zip code, then to extract, but there are many mistakes in the compilation process, I think this is caused by the update of MOOSE in recent years (I am currently using MOOSE version 2024.11.21), I would like to ask if there is a suitable Crow that you can provide. Thank you very much and look forward to hearing from you.

**Reply by GiudGiud** [reply]
You should see if there are any working forks

https://github.com/SudiptaBiswas/Crow/forks?include=active&page=1&period=&sort_by=stargazer_counts

otherwise the update to making it compile is not very hard. it s just this one validParams item you need to fix
https://mooseframework.inl.gov/newsletter/2020/2020_04.html

**Reply by HPL001** [reply]
@GiudGiud Thank you very much for your reply, which is of great help to me. I downloaded Crow from 3 years ago（https://github.com/luchinsky/Crow） and MOOSE from 3 years ago（https://github.com/yardasol/moose）, but I had some problems compiling. Please forgive me for not being familiar with the compilation process, the following is a screenshot of the problem, I hope to get your guidance. Thank you very much and look forward to hearing from you.
<img width="879" alt="001" src="https://github.com/user-attachments/assets/409db41d-cca4-4d93-aa60-32d7acdcc242">

**Reply by GiudGiud** [reply]
Are you using the conda packages from 3 years ago as well?

**Reply by HPL001** [reply]
Hello, I replaced the conda 3 years ago according to your suggestion. However, I have a question about whether I should choose to use the three year old moose environment here, and if so, what command I should use when I create the environment in the tutorial on installing moose on the installation website (conda create -n moose moose). Looking forward to hearing from you. Thank you very much.

**Reply by GiudGiud** [reply]
I actually don’t think we offer these conda packages still. 

I would update moose and the conda packages to the current ones and fix any potential breakage in crow

**Reply by HPL001** [reply]
It would be great if you would like to update Crow to suit the latest moose and conda versions. Thank you very much and look forward to your work.

**Reply by GiudGiud** [reply]
By "I would" I mean this is what I would do if I were you.
There is too much activity right now for me to do it

**Reply by HPL001** [reply]
Thank you for your advice, but unfortunately I am not very familiar with MOOSE, which may take a lot of time, so I would like to ask how to install the MOOSE environment back to three years ago?Can I use this line command "conda create -n moose moose-dev=2024.11.15=mpich" and specify the version? How do I find the release number?

**Reply by GiudGiud** [reply]
you can see the list here
https://conda.software.inl.gov/public/osx-arm64/
oldest you can download for arm-macos is August 2022 which is too late (considering the old validParams removal post-deprecation was in early 2021)

**Reply by HPL001** [reply]
Thanks for replying, I am a linux system environment. I saw on the website (https://conda.software.inl.gov/public/linux-64/) have a version（[moose-2021_07_28-build_0.tar.bz2](https://conda.software.inl.gov/public/linux-64/moose-2021_07_28-build_0.tar.bz2)）, don't know if I can use? If so, what commands should I use to download and configure the conda environment?

**Reply by GiudGiud** [reply]
It might work, the removal of support for the feature is February 2022
https://mooseframework.inl.gov/newsletter/2022/2022_02.html
You can try doing this:
```
conda install moose-libmesh=2021.07.28
```
but at this point you are on your own using MOOSE from 3 years ago. If you hit any difficulty, you need to update

**Reply by GiudGiud** [reply]
@milljm would you know the command for this old package?

**Reply by milljm** [reply]
The `moose-2021-*` package is an actual moose binary package (not very useful for development). And another problem with that package is that they usually require the other supplemental packages (like moose-libmesh-2021-* of the same date). And those packages are gone. The earliest support packages begin at 2022-* timeframe.

If ultimately required, all of these packages can be regenerated via the Conda recipes in your MOOSE repo, from whatever time fame you need, in `moose/conda/*`. The directories in this location make up all the Conda packages that end up on our public channel. I can walk you through the process if you so desire. But be warned: building these packages will require an absolute clean and sane environment (no HPC modules loaded, no Homebrew, etc.

**Reply by milljm** [reply]
Looks like I should remove that package, since its uninstallable...

**Reply by HPL001** [reply]
@milljm Thanks for your reply. I am not very familiar with MOOSE, so it is difficult for me to update Crow to the existing version, so I prefer to use the previous version of MOOSE. So what should I do to get a three year old version, please guide me. I now have a new linux system and conda.

**Reply by milljm** [reply]
There might be one possible way around this. We do supply an archive channel. Though, the very reason we created it was because of some catastrophe that saw many of the package dependencies break. 

Try this:

```bash
conda create -n oldstack moose-libmesh=2021.07.14 --channel https://conda.software.inl.gov/archive
conda activate oldstack
```

**Reply by milljm** [reply]
what does the following report?

```bash
conda search moose-libmesh --channel https://conda.software.inl.gov/archive
```

**Reply by milljm** [reply]
Hmm, no. That channel is very much active and available. I am able to reach it from work, and from my personal internet supplier.

What happens when you try to go there manually:

https://conda.software.inl.gov/archive

**Reply by milljm** [reply]
Are you operating behind a corporate proxy? Can you curl the address from the same terminal?

```bash
curl https://conda.software.inl.gov/archive
```
should respond with:
```pre
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="https://conda.software.inl.gov/archive/">here</a>.</p>
</body></html>
```

Edit: Sorry for the delayed responses... I don't seem to be being notified when you post.

**Reply by milljm** [reply]
Its odd that `https://conda.software.inl.gov/public` would work, but not `/archive`. I would say this is something to do with proxies, or certificates (like we have to do at a gov lab):

```pre
REQUESTS_CA_BUNDLE=/opt/certs/CAINLROOT_B64.crt
SSL_CERT_FILE=/opt/certs/CAINLROOT_B64.crt
CURL_CA_BUNDLE=/opt/certs/CAINLROOT_B64.crt
```

But these things control how domains are accessed, not a trailing _directory_ in the URL...

My advice is to check the settings in your favorite web browser you used to click on the link that works, see how _its_ configured to operate (proxies), and try to mimic those on the command line:

```bash
export http_proxy=<whatever firefox says>
export https_proxy=<whatever firefox says>
```
...etc

**Reply by milljm** [reply]
Or, if it becomes a hassle to deal with the network issue, we can instead try to build the necessary packages manually.

**Reply by GiudGiud** [reply]
maybe the manual install (using the scripts) is easier than building the conda packages
this way
https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html

**Reply by HPL001** [reply]
Thank you very much for your guidance, if I follow the steps in the link to install, do I need to specify the corresponding version? (Because I've had some issues with Crow compilation in the past that may be related to the compiler version, but I'm not really sure)

**Reply by milljm** [reply]
Actually the method you pointed out would be much easier! Lets see how far we get.

Install the compiler stack (and clean the moose repo thoroughly):
```bash
conda create -n compilers moose-mpi mpich python=3.10
conda activate compilers
cd ~/projects/moose

# before running the next command, be sure that you have saved any
# uncommitted work. Because we're about to delete stuff
git clean -xfd; git submodule foreach --recursive git clean -xfd
```
Build PETSc:
```bash
./scripts/update_and_rebuild_petsc.sh
```
If PETSc builds okay, move on to libMesh:
```bash
./scripts/update_and_rebuild_libmesh.sh
```
If libMesh builds okay, we can now attempt to build MOOSE and/or your application. Though, like what we did for MOOSE, we should probably do a deep clean on Crow (or your application). And like what was stated for moose, you should commit any changes not being tracked, as they will be deleted by running the following commands:

```bash
cd ~/projects/crow   # or your app
git clean -xfd; git submodule foreach --recursive git clean -xfd
```

Once we're in a clean state, try to build again...

PS: I have not read through this entire thread, so I am not really sure if we are attempting to build Crow, run Crow, etc...

**Reply by HPL001** [reply]
Thank you very much for your patient guidance, I will follow your steps, but because I used the previous version of conda, the process may be a little slow, and then I will timely feedback the results to you (no later than tomorrow). You mentioned the new steps with the aforementioned link (https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html) have what relation? Do I just need to follow the new steps you provided on github? PS: I work with a new miniforge(3 years gao version).

**Reply by HPL001** [reply]
Hello, I noticed there is a step here(cd ~/projects/moose and cd ~/projects/crow), do I need to install moose in advance? How do I install moose in this step if needed? (Because I want to get moose in 2021 and the corresponding environment, dependency packages)

**Reply by milljm** [reply]
I am assuming you had MOOSE and Crow at the version you already require. I don't know anything about Crow... does Crow have a moose submodule? That would be ideal, as then Crow it self would control what version of MOOSE _it_ requires at the time.

Other than that, I don't have a HASH off the top of my head for a period of MOOSE back in 2021 that you're after. If there is no MOOSE submodule in Crow, you'll instead need to do something like:

```bash
# enter into the moose directory, where ever that is
cd moose

# figure out all the available hashes around this timeframe
git log | grep -B 3 "Jul 14" | grep -C 3 "2021"
```
A lot of commits will return. You'll want to simply pick one of hashes, and go from there

```bash
git checkout abcdefg12345
cd scripts
./update_and_rebuild...    # as mentioned in the earlier post
```

**Reply by milljm** [reply]
Is there a way you can post text as text instead of images of text? I can't for example, easily clone the version of PETSc you are grabbing... because I can't select text from a picture.

What version of MOOSE did you end up cloning? Perhaps that version of MOOSE is too old to make use of newer compilers. I'd like to try for myself...

**Reply by HPL001** [reply]
I cloned this version of moose (https://github.com/yardasol/moose.git), and this version of Crow (https://github.com/luchinsky/Crow.git). Some of the information in the figure is in the txt file.
[001.txt](https://github.com/user-attachments/files/17907766/001.txt)

**Reply by milljm** [reply]
I need you to enter the MOOSE directory and run:
```bash
git rev-parse HEAD
```

This will tell me what hash you checked out.

**Reply by HPL001** [reply]
ok, this is the output.
[002.txt](https://github.com/user-attachments/files/17912562/002.txt)

**Reply by HPL001** [reply]
@milljm Hello, dear developer. Are there any solutions now? Looking forward to your reply.
